### PR TITLE
Added unit tests for AudioPlayer

### PR DESCRIPTION
**Pull request by TheItsyBitsySpider**
## Purpose / Description
Added unit tests that cover all of the functions found in AudioPlayer. According to the coverage report generated by Android Studio, it should be at 100% coverage now.

Also in testAbruptStop, is it intended behavior for stoppingListener and stoppedListener to not run() when AudioPlayer plays, then immediately stops with stop()?

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/test/java/com/ichi2/anki/multimediacard/AudioPlayerTest.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+        ShadowMediaPlayer testPlayer = new ShadowMediaPlayer();
+        DataSource mFileSource = DataSource.toDataSource(mFile.getAbsolutePath());
+        ShadowMediaPlayer.MediaInfo mFileInfo = new ShadowMediaPlayer.MediaInfo();
+        testPlayer.addMediaInfo(mFileSource, mFileInfo);
+    }
+
+    @Rule
+    public TemporaryFolder tempFolder = new TemporaryFolder();
+
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/anki/multimediacard/AudioPlayerTest.java:44]: We generally put class members on top of methods. Just a nitpick

```suggestion
    @Rule
    public TemporaryFolder tempFolder = new TemporaryFolder();

    @Before
    public void before() throws IOException {
        mAudioPlayer = new AudioPlayer();
        mFile = tempFolder.newFile("testaudio.wav");

        ShadowMediaPlayer testPlayer = new ShadowMediaPlayer();
        DataSource mFileSource = DataSource.toDataSource(mFile.getAbsolutePath());
        ShadowMediaPlayer.MediaInfo mFileInfo = new ShadowMediaPlayer.MediaInfo();
        testPlayer.addMediaInfo(mFileSource, mFileInfo);
    }

```

**Review by mikehardy — CHANGES_REQUESTED** [review]
Thanks for this!

> Also in testAbruptStop, is it intended behavior for stoppingListener and stoppedListener to not run() when AudioPlayer plays, then immediately stops with stop()?

Honestly, I'm not sure! One of the things testing points out is odd corner cases, and then (as is likely the case here) the person doing the test has to go investigate them and then they are the expert and either fix an issue in the code / cementing it with the investigated assertion in the test, or assert strongly in a comment that this was checked (maybe with a link to an authoritative source) and it's correct

I certainly couldn't say if the listeners are supposed to fire if there is an abrupt stop. Seems like they maybe should be ? But they aren't...so, I don't know

**Comment by TheItsyBitsySpider** [comment]
After looking at AudioPlayer and the Android MediaPlayer class, the stop listeners are connected to OnCompletionListener, so it activates only if it reaches the end during playback and not when abruptly stopped.

It's working as coded, but I don't have a thorough enough understanding of the whole system to make changes. Perhaps it's just a naming clarity issue and the AudioPlayer class's mOnStoppingListener and mOnStoppedListener should be renamed to reference CompletionListener, but since there's no complaints about the current behavior, I'm not going to be looking into it further at this point.

File: AnkiDroid/src/test/java/com/ichi2/anki/multimediacard/AudioPlayerTest.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        Runnable stoppedListener = mock(Runnable.class);
+
+        mAudioPlayer.setOnStoppingListener(stoppingListener);
+        mAudioPlayer.setOnStoppedListener(stoppedListener);
+
+        runTasksInBackground();
+        mAudioPlayer.play(mFile.getAbsolutePath());
+        mAudioPlayer.stop();
+        advanceRobolectricLooper();
+        //audio has immediately stopped, and the stop listeners should not have run
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/anki/multimediacard/AudioPlayerTest.java:118]:
```suggestion
        // audio was stopped prior to completion, and on investigation of the "stop" listeners,
        // it appears they are hooked to the *completion* of the audio, so it appears correct that
        // no "stop" listeners were called as the audio did not in fact complete
```

**Review by mikehardy — APPROVED** [review]
Sounds good to me, and I mean this sincerely - the deeper investigation is really valuable
So valuable I took that hard work and put it in the comment around the test expectation,
in the future this may serve to help future developers understand the behavior in case
the assumption is either wrong or just poorly understand by the future person, which could
be key to getting something done in the future.

Thanks!
